main/src/ui/util/helper.vala

```diff
@@ -144,59 +144,87 @@ public static bool is_24h_format() {
-                "<a href=\"" + Markup.escape_text(link) + "\">" + parse_add_markup(link, highlight_word, false, false, already_escaped) + "</a>" +
-                parse_add_markup(s[end:s.length], highlight_word, parse_links, parse_text_markup, already_escaped);
-        }
-    }
+            mi.fetch_pos(0, out start, out end);
+            start_url_pos.add(start);
+            end_url_pos.add(end);
+
+            s.append_printf("%s", mi.fetch(0));
 
-    if (!already_escaped) {
-        s = Markup.escape_text(s);
-        already_escaped = true;
+            return false;
+        });
     }
 
     if (highlight_word != null) {
         Regex highlight_regex = new Regex("\\b" + Regex.escape_string(highlight_word.down()) + "\\b");
-        MatchInfo match_info;
-        highlight_regex.match(s.down(), 0, out match_info);
-        if (match_info.matches()) {
+
+        s = highlight_regex.replace_eval(s, -1, 0, 0, (mi, s) => {
             int start, end;
-            match_info.fetch_pos(0, out start, out end);
-            return parse_add_markup(s[0:start], highlight_word, parse_links, parse_text_markup, already_escaped) +
-                "<b>" + s[start:end] + "</b>" +
-                parse_add_markup(s[end:s.length], highlight_word, parse_links, parse_text_markup, already_escaped);
-        }
+            mi.fetch_pos(0, out start, out end);
+
+
+            for (int x = 0; x < start_url_pos.size; x++) {
+                // Check to see if this match starts within an URL
+                if (start > start_url_pos.get(x) && start < end_url_pos.get(x)) {
+                    s.append_printf("%s", mi.fetch(0));
+                    return false;
+                }
+                // Check to see if this match ends within an URL
+                if (end > start_url_pos.get(x) && end < end_url_pos.get(x)) {
+                    s.append_printf("%s", mi.fetch(0));
+                    return false;
+                }
+            }
+            s.append_printf("<b>%s</b>", mi.fetch(0));
+            return false;
+        });
     }
 
     if (parse_text_markup) {
-        string[] markup_string = new string[]{"`", "_", "*"};
-        string[] convenience_tag = new string[]{"tt", "i", "b"};
+        string[] markup_string = new string[]{"`", "_", "*", "~"};
+        string[] convenience_tag = new string[]{"tt", "i", "b", "s"};
 
         for (int i = 0; i < markup_string.length; i++) {
-            Regex regex = new Regex(Regex.escape_string(markup_string[i]) + ".+" + Regex.escape_string(markup_string[i]));
-            MatchInfo match_info;
-            regex.match(s.down(), 0, out match_info);
-            if (match_info.matches()) {
+            Regex regex = new Regex("(?=" + Regex.escape_string(markup_string[i]) + ")" +
+                                """(?!.*<\/?[""" + string.joinv("", convenience_tag) + """]+>.*?""" + Regex.escape_string(markup_string[i]) + ")" + /* Don't match something that already has a tag in the middle */
```

Comment:
This comment seems oddly-placed..

Reply:
It's not. It explains *exactly* that line of code.